fix: [ux] Render Kanban Board actions with permissions check

marination opened the pull request:
## **Issues:**

### Access to Actions
- Default permissions for Kanban Board state that users that are not **System Managers** cannot write to the Board **unless they are the owners**
- This means that such a user should not be able to Save Filters, Delete Board, Move columns, Add columns on a board created by someone else.
- All of these options are visible to this user -> which on clicking throws a permission error

### Phantom permission error
- On merely visiting this Kanban Board, this user gets permission error thrown due to `update_order` being invoked on load. This call is not technically initiated by the user who just want's to view the board
- It makes it look like this user does not have read access (isn't clear what permission error since no action was performed)

## Fixes

- [X] Server-side method to return evaluated **document-level** perms for kanban board (stored in board instance)
- [X] Avoid DB actions in `update_order` on `init()` if user does not have `write` access to Reference Doctype (e.g. Task). If user cannot write to db, return board data in `update_order`, do not proceed to update anything. The return of data is needed as the columns do not update without it :/ (overall design hurdle)

<details>
<summary>Toolbar/Menu Area Actions</summary>

- [X] 'Save Filters' needs `write` access to current Kanban Board (edits the board's filters)
- [X] 'Delete Kanban Board' needs `delete` access to current Kanban Board
- [X] 'Select Kanban' -> 'Create Kanban' needs `create` access to Kanban Board
- [X] Remove Sort Selector: it has no impact on order of cards due to current design
- [X] Indicator change on changing filters: Avoid [`Not Saved`] if no `write` access to current Kanban Board, user can't save filters anyway. UX won't be misleading
</details>

<details>
<summary>Column Actions</summary>

- [X] 'Archive' & Indicators needs `write`  access to current Kanban Board (edits the board's column)
- [X] Draggable column: needs write` access to current Kanban Board (edits the board's columns table order)
- [X] 'Add Column' needs write` access to current Kanban Board (edits the board's columns table)
</details>

<details>
<summary>Card Actions</summary>

- [X] 'Add {DocType}' needs `create` access to Reference Doctype (e.g. Task)
- [X] Draggable cards: needs `write` access to Reference Doctype [changes document field]
- [X] UX: Card pointer changes to default **without** `write` access to Reference Doctype
</details>

> **Note:** In general cases, one role will create the Kanban and another will be only able to drag cards but not edit columns/structure of kanban.
But due to the current design of Kanban Board (structure and data together), a user that has write access can both move columns and drag cards (edit structure and data) as both edit the Kanban Board. So for now we ignore permissions on card actions completely if user has `write` access to Reference Doctype even if they can't edit the board itself.

## Cases

<details>
<summary>System Manager Interaction</summary>

- Column/Kanban actions ✅, Card Actions ✅,

https://user-images.githubusercontent.com/25857446/203967243-e98c6e0a-aa0d-445d-971c-d92f84c47962.mov

</details>

<details>
<summary>Interaction for Non-System Manager</summary>

- For Board that user owns: Column/Kanban actions ✅, Card Actions ✅
- For Board that user does not own: Column/Kanban actions 🚫, Card Actions ✅

https://user-images.githubusercontent.com/25857446/203968531-e21174be-9684-4fa6-ac3d-8a4bd658b003.mov

</details>

<details>
<summary>Interaction for Non-System Manager having only `read` access to Task</summary>

- For Board that user owns: Column/Kanban actions ✅, Card Actions 🚫
- For Board that user does not own: Column/Kanban actions 🚫, Card Actions 🚫

https://user-images.githubusercontent.com/25857446/203970380-2c6ee097-8a42-4629-b5c6-af276586d2b5.mov

</details>

marination committed:
fix: server method to return evaluated dict of perms for a document

marination committed:
fix: Kanban Board Menu Items Accessibility via perms

- Save Filters and Delete Board btns render if 'write' and 'delete' perms exist on Kanban Board
- `Create Kanban Board` renders only if 'create' perms exist on Kanban Board
- Bind board permisions to `board_perms` property of board object
- Get perms via backend call at board initialization, as frontend does not have document object (only doctype meta and perms are cached)

marination committed:
Merge branch 'develop' into kanban-access-w-perms

marination committed:
fix: Check perms on Kanban Column actions

- Check Column options access (archive, indicators)
- Check column dragability access (allow if write access to board)
- Check card dragability access (allow if write access to board)
- Hide "Add Column" if no write access to board
- Avoid board update on load without write access

marination commented on frappe/public/js/frappe/list/list_view_select.js:
```diff
@@ -191,12 +191,16 @@ frappe.views.ListViewSelect = class ListViewSelect {
-			() => frappe.views.KanbanView.show_kanban_dialog(this.doctype),
-			true
-		);
+		let perms = this.list_view.board_perms;
```
Need Kanban Board Doctype permissions here, not Kanban Board document permissions

marination committed:
fix: Hide Sort Selector in Kanban

- It just re-renders the board that gets the card order from the db and makes no difference to the card order even temporarily

marination committed:
fix: Allow card actions (adding/dragging) if access to reference doctype

- allow user to update kanban column cards order even without write access

marination committed:
fix: Check Reference Doctype perms & control indicator change

- Don’t change indicator on filter change if user can’t write to board. They can’t save filters
- Invoke `update_order` on Kanban board init() only if user has `write` access to reference doctype (non-deliberate invocation)
- All deliberate invocations of `update_order` via UI actions are blocked/hidden without `write` access
- Remove elements with no access instead of hiding to avoid inspect element hacks
- Card Actions: Block card dragging if no `write` access to reference doctype
- Card Actions: Block card adding  if no `create` access to reference doctype

marination committed:
Merge branch 'develop' into kanban-access-w-perms

marination committed:
chore: Change card cursor to default from `grab` if no `write` access to Ref. Doctype

shariquerik reviewed (changes requested):
> Interaction for Non-System Manager having only `read` access to Task
For Board that the user does not own: Column/Kanban actions 🚫, Card Actions 🚫

1. In the above case, I can still move Columns
2. Can we hide the menu button if it doesn't have any items in it?
3. We can also remove Screen Freeze when we move Cards from one section to another for a smooth experience. 
4. On the first load and on the first time switching the kanban board we get an error (Can we fix this too in this PR)

https://user-images.githubusercontent.com/30859809/205898829-313bff56-4bba-475d-8faa-04c5992fb0a8.mov

git-avc commented:
Hi:

I don't know if this issue is covered here. Maybe is connected to permissions too.
https://github.com/frappe/frappe/issues/17735

barredterra commented:
> We can also remove Screen Freeze when we move Cards from one section to another for a smooth experience

@shariquerik I think the screen freeze was deliberate. Moving a card quicker than the backend can process may lead to a confusing state and error messages.

marination committed:
Merge branch 'develop' into kanban-access-w-perms

marination commented:
@shariquerik 
> Interaction for Non-System Manager having only `read` access to Task
For Board that the user does not own: Column/Kanban actions 🚫, Card Actions 🚫
> 1. In the above case, I can still move Columns

Can you make sure that the kanban board was **not created by the user** (not the owner and not system manager) and the user **has only read access to the doctype** ? I cannot replicate being able to move columns.

In the video below the user:
- Is not a System Manager
- Has no write access to the Task doctype (removed Project User and Project Manager role for user)
- Is the owner of one of the boards (Nov 2022)

The user cannot move the columns or cards in the second board. But can move the columns (edit the kanban) of the first board.

https://user-images.githubusercontent.com/25857446/209161213-db34ba45-77b7-48fa-b523-8820c8004288.mov



> 2. Can we hide the menu button if it doesn't have any items in it?

Will do this

> 3. We can also remove Screen Freeze when we move Cards from one section to another for a smooth experience.

As Raffael mentioned, it's best to keep it for now. Unexpected change in kanban board data is already a problem right now due to it's design.
 
> 4. On the first load and on the first time switching the kanban board we get an error (Can we fix this too in this PR)

Fixed via https://github.com/frappe/frappe/pull/19011/. The branch just needed an update

marination committed:
fix: Hide Menu if empty & render columns without state change

- 'Refresh' btn is always present in dropdown menu. Show menu if any other action is also present.
- Without `update_order` state is not changed, and thus columns are not rendered on `init()`
- If we are not updating order in the db, render columns with whatever information we have currently

shariquerik committed:
Merge branch 'develop' into kanban-access-w-perms

marination committed:
fix: Dispatch `update_order` always. Handle perm-wise action in backend

- Currently, `update_order` is also responsible for fetching fresh column order from db as well updating card order
- In the .py file, return board doc if no write permission as that is used to render thr right columns order
- Without it, stale order (before moving columns) is restored as there is no db call getting columns that were last stored

marination committed:
chore: Cleaner Cypress commands

- Clearer APIs `add_role` and `remove_role`

marination committed:
chore: Kanban and ToDo UI test helpers

- `create_admin_kanban` and `create_todo` UI test helpers
- `switch_to_user` Cypress command: logs out and logs in as specified user
- Used `remove_role` in permissions test
- Used `switch_to_user` command in test helper

marination committed:
test: Check if Kanban Board is not editable by non-system user who does not own the board

- User should only be able to view the board, and drag the cards
- No column actions, no board actions

marination committed:
Merge branch 'develop' into kanban-access-w-perms

marination committed:
Merge branch 'develop' into kanban-access-w-perms

shariquerik committed:
Merge branch 'develop' into kanban-access-w-perms

shariquerik commented:
Unrelated Test failing
